scapy/contrib/isotp/isotp_native_socket.py

```diff
@@ -319,44 +319,63 @@ def __init__(self,
+        # store arguments internally
         self.tx_id = tx_id
         self.rx_id = rx_id
         self.ext_address = ext_address
         self.rx_ext_address = rx_ext_address
-
-        self.can_socket.setsockopt(SOL_CAN_ISOTP,
-                                   CAN_ISOTP_RECV_FC,
-                                   self.__build_can_isotp_fc_options(
-                                       stmin=stmin, bs=bs))
-        self.can_socket.setsockopt(SOL_CAN_ISOTP,
-                                   CAN_ISOTP_LL_OPTS,
-                                   self.__build_can_isotp_ll_options(
-                                       mtu=CAN_ISOTP_CANFD_MTU if fd
-                                       else CAN_ISOTP_DEFAULT_LL_MTU,
-                                       tx_dl=CAN_FD_ISOTP_DEFAULT_LL_TX_DL if fd
-                                       else CAN_ISOTP_DEFAULT_LL_TX_DL))
-        self.can_socket.setsockopt(
+        self.bs = bs
+        self.stmin = stmin
+        self.padding = padding
+        self.listen_only = listen_only
+        self.frame_txtime = frame_txtime
+        self.fd = fd
+        if basecls is None:
+            log_isotp.warning('Provide a basecls ')
+        self.basecls = basecls
+        self._init_socket()
+
+    def _init_socket(self) -> None:
+        can_socket = socket.socket(socket.PF_CAN, socket.SOCK_DGRAM,
+                                   CAN_ISOTP)
+        self.__set_option_flags(can_socket,
+                                self.ext_address,
+                                self.rx_ext_address,
+                                self.listen_only,
+                                self.padding,
+                                self.frame_txtime)
+
+        can_socket.setsockopt(SOL_CAN_ISOTP,
+                              CAN_ISOTP_RECV_FC,
+                              self.__build_can_isotp_fc_options(
+                                  stmin=self.stmin, bs=self.bs))
+        can_socket.setsockopt(SOL_CAN_ISOTP,
+                              CAN_ISOTP_LL_OPTS,
+                              self.__build_can_isotp_ll_options(
+                                  mtu=CAN_ISOTP_CANFD_MTU if self.fd
+                                  else CAN_ISOTP_DEFAULT_LL_MTU,
+                                  tx_dl=CAN_FD_ISOTP_DEFAULT_LL_TX_DL if self.fd
+                                  else CAN_ISOTP_DEFAULT_LL_TX_DL))
+        can_socket.setsockopt(
             socket.SOL_SOCKET,
             SO_TIMESTAMPNS,
             1
         )
 
-        self.__bind_socket(self.can_socket, self.iface, tx_id, rx_id)
-        self.ins = self.can_socket
-        self.outs = self.can_socket
-        if basecls is None:
-            log_isotp.warning('Provide a basecls ')
-        self.basecls = basecls
+        self.__bind_socket(can_socket, self.iface, self.tx_id, self.rx_id)
+        # make sure existing sockets are closed
```

Comment:
Hm.
I dislike having duplicate code. This is a duplicate of close, can't you just use it?

Reply:
reusing close(), now!